### ISSUE TITLE
.env file affects tests

### ISSUE THREAD

**Issue by mejlholm** [quarkusio/quarkus#11438]
**Describe the bug**
The .env file is a great way to set environment variables, which should not go into source control. It should however only work for the dev profile, not the test profile, as this will lead to flaky tests that might work on one machine, but fails in the CI. 

**Expected behavior**
Test cases are unaffected by the .env file. 

**Actual behavior**
Test fails, because the .env file sets a value, which overwrites the value provided in application.properties. 

**To Reproduce**
Check this reproducer:
https://github.com/mejlholm/env-file-test-reproducer

**Configuration**
Not relevant

**Screenshots**
Not relevant

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
MSYS_NT-10.0-19041 SNBWS0027 3.1.6-340.x86_64 2020-07-09 14:33 UTC x86_64 Msys
Microsoft Windows [Version 10.0.19041.388]

 - Output of `java -version`: 
java -version
java version "1.8.0_261"
Java(TM) SE Runtime Environment (build 1.8.0_261-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.261-b12, mixed mode)

 - Quarkus version or git rev: 
1.7.0.Final

 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: C:\dev\tools\apache-maven-3.6.3\bin\..
Java version: 14.0.1, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk-14.0.1
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

**Additional context**
Not relevant

**Comment by famod** [comment]
This was added explicitily to `io.quarkus.runtime.configuration.ConfigUtils` via #7530.

AFAICS, https://quarkus.io/guides/config#overriding-properties-at-runtime _does not limit_ `.env` to dev mode.

At this point I'd be inclined to close this issue as "works as designed" but I understand that one might want to use individual dev mode only settings locally.
I haven't tried myself, but did you try to use the dev profile in `.env`, @mejlholm?

/cc @geoand

**Comment by geoand** [comment]
Thanks @famod.

Yes, this does work as expected, .env isn't meant to work only for dev mode, it works the same in all modes.

I can understand why someone would want to limit its use, but then I am pretty sure that other folks would then want to have it back 😉.

Moreover, . env is familiar from other language environments as a sort of alternative to setting environment variables, therefore if we did limit its use to only dev mode, the experience would not be in line with other environments.

One way to work around this would be to have confirmation in your build tool that effectively renames the file before tests are run and reverts the change once tests are finished.

In conclusion, I'm gonna close this as works as expected.

**Comment by machi1990** [comment]
Solid points @famod @geoand 

Just to add on this.

 > One way to work around this would be to have confirmation in your build tool that effectively renames the file before tests are run and reverts the change once tests are finished.

+1, in most projects the `.env` file is never to be committed for the sake of example to onboard new team members, an alternative file `.env.example` file can be committed containing fake values for each configuration file.

**Comment by mejlholm** [comment]
Can't really say that i follow the usage pattern here. If the .env file is to apply to both dev and test modes, why not just put it all in application.properties? 

Generally we put everything we can into applications.properties, unless it is a secret. In dev mode we need a way to values for secrets into the application, which is where the .env file is handy. For test values, it is fine to provide a %test value in application.properties. 

I've checked that dev profile in .env works, so that will probabaly be the solution for us.

**Comment by famod** [comment]
@mejlholm 
> I've checked that dev profile in .env works, so that will probabaly be the solution for us.

Ok, cool. Never used a profile in `.env`. How does it look like? `%DEV_HELLO_VALUE=other`?

**Comment by mejlholm** [comment]
Just tried %dev.HELLO_VALUE.

**Comment by famod** [comment]
Ah, I see - same syntax as in regular properties. Makes sense. Thanks!

**Comment by gaetancollaud** [comment]
> Just tried %dev.HELLO_VALUE=other
> Seems to work.

This didn't work for me.

The actual syntax is `_DEV_HELLO_VALUE`. See documentation here: https://quarkus.io/guides/config-reference#profile-in-the-property-name

**Comment by geoand** [comment]
This was brought to my attention again by @rvansa.

Maybe we could have some arcane system property that disables `DotEnvConfigSourceProvider`? @radcortez  WDYT?

**Comment by radcortez** [comment]
Not sure if this is a good idea. I understand the frustration, but this may happen in other scenarios (for instance `application.properties` in the `config` folder).  Unless we define a specific rule for test mode, we may end up with some inconsistency.

My recommendation is to just use specific profile `env` files (`env-prod`, `env-dev`, etc).

**Comment by geoand** [comment]
I like your per profile file a lot.

I'll have a look

**Comment by geoand** [comment]
I looked into this a little and it seems like per profile `.env` files are implicitly supported already.

The only caveat is that you need to have an `.env` file as well (even if it's empty). @radcortez this happens because https://github.com/smallrye/smallrye-config/blob/main/implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceLoader.java#L110-L111 is only executed if the original file exists. Would it make sense to relax this constraint?

**Comment by radcortez** [comment]
> I looked into this a little and it seems like per profile `.env` files are implicitly supported already.

Yes, this is already supported. Sorry if I was not clear on that.

> The only caveat is that you need to have an `.env` file as well (even if it's empty). @radcortez this happens because https://github.com/smallrye/smallrye-config/blob/main/implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceLoader.java#L110-L111 is only executed if the original file exists. Would it make sense to relax this constraint?

The reason why the constraint exists is to be able to disambiguate which is the main file and which are profile files. This is required to keep a consistent loading order with classpath resources.

In the case of `.env`, maybe we can lift that restriction, but it will become inconsistent with the other case.

**Comment by geoand** [comment]
> In the case of .env, maybe we can lift that restriction, but it will become inconsistent with the other case.

Okay, then I vote for being consistent

**Comment by rvansa** [comment]
In my case the `.env` is written to support dev environment - podman-compose starts some containers which have the cwd mounted and write down secrets into the `.env` file. I could prefix all the properties in the file with `_DEV_` to let test ignore them.

**Comment by geoand** [comment]
You could have an empty `.env` file and then have a `.env-dev` with the properties you want applied for dev-mode only
